4.x form protector

ADmad opened the pull request:
Currently the code form tampering protection token generation and verification is split across `SecurityComponent` and `FormHelper`.  I created a new `FormProtector` class which now encapsulates all related code.

`FormHelper` has been updated to delegate all form token generated related stuff to the new class.

For token verification I don't want to try and retrofit the `SecurityComponent` to use the new class. Two possible are we either create a new `FormProtectionComponent` or a new `FormProtectionMiddleware` and simply deprecate `SecurityComponent`.

The benefit of creating a component is configuring unlocked field and turning of POST validation on per action basis becomes convenient.

TODO:

- [x] Create `FormProtectionComponent` or `FormProtectionMiddleware`.
- [x] Remove / deprecate `SecureFieldTokenTrait` trait.

ADmad committed:
Add FormProtector class.

ADmad committed:
Move form protection token generation from FromHelper to FormProtector.

ADmad committed:
Cleanup FormProtector.

ADmad committed:
Change visibility and removed unused trait.

ADmad committed:
Fix errors reported by psalm.

markstory commented on src/Form/FormProtector.php:
```diff
@@ -0,0 +1,563 @@
+
+use Cake\Core\Configure;
+use Cake\Utility\Hash;
+use Cake\Utility\Security;
```
Does this add new dependencies to cake/form?

dereuromark replied:
Nope, as validation https://github.com/cakephp/validation/blob/master/composer.json#L27 which already is required.

ADmad replied:
As @dereuromark pointed out it doesn't. Though I doubt anyone uses the `Form` package as standalone anyway.

markstory commented on src/Form/FormProtector.php:
```diff
@@ -0,0 +1,563 @@
+ * - Form's action (URL) is not modified.
+ * - Unknown / extra fields are not added to the form.
+ * - Existing fields have not been removed from the form.
+ * - Values of hidden inputs have not been changed.
```
Should this class be marked `@internal`?

ADmad replied:
Done

markstory commented on tests/TestCase/View/Helper/FormHelperTest.php:
```diff
@@ -266,15 +278,18 @@ public function testOrderForRenderingWidgetAndFetchingSecureFields()
         $result = $this->Form->widget('select', ['secure' => true, 'name' => '0']);
         $this->assertSame('<select name="0"></select>', $result);
-        $this->assertEquals(['0'], $this->Form->fields);
+        $result = $this->getProtectedProperty($this->Form->getFormProtector(), 'fields');
```
Should FormProtector provide get methods for inspecting its state? That would save us using reflection.

ADmad replied:
Except for tests I don't see a need to have getters for the properties. Perhaps just adding `__debugInfo()` would suffice?

markstory replied:
Yeah, pulling out of `__debugInfo()` is reasonable for tests to me.

ADmad committed:
Move field name string to array conversion into FormProtector.

ADmad committed:
Add FormProtector::__debugInfo() to avoid reflection in tests.

dereuromark commented:
I dont know how relevant this is, but the old FormHelper you added to Shim 4.x branch, seems to fail with changes to 4.x now:
https://travis-ci.org/dereuromark/cakephp-shim/jobs/595381701#L317

ADmad commented:
@dereuromark I'll add tests for 3.x style multiple selects for datetime and check.

ADmad commented:
I checked and the tests for the Shim plugins FormHelper will have to be updated as the property it's using in tests doesn't exist anymore. There won't be any actual issues in using the datetime selects.

ADmad commented:
So for the form token verification should I add a `FormProtectionMiddleware` or `FormProtectionComponent`?

markstory commented:
@ADmad I think a component makes sense here for the reasons you outlined around unlockedFields and disabling per action.

ADmad committed:
Add FormProtectionComponent.

ADmad committed:
Rename request attribute "formToken" to "formTokenData".

Previous name was misleading. It does not contain the form token but rather
the data used to the generate token.

ADmad committed:
Update docblock.

ADmad committed:
Deprecate SecurityComponent.

ADmad committed:
Remove SecureFieldTokenTrait.

It's functionality has been moved into FormProtector.

ADmad commented:
- Added `FormProtectionComponent`
- Removed `SecureFieldTokenTrait`.
- Deprecated `SecurityComponent`.

ADmad commented:
`SecurityComponent` does some funky juggling related to which expection it throws and with what message.

I simplified `FormProtectionComponent` to always throw `BadRequestException` and just use more detailed error message if debug is turned on.

markstory commented on src/Controller/Component/FormProtectionComponent.php:
```diff
@@ -0,0 +1,162 @@
+            throw new BadRequestException(static::DEFAULT_EXCEPTION_MESSAGE);
+        }
+
+        $errorMessage = $formProtector->getError() ?: static::DEFAULT_EXCEPTION_MESSAGE;
```
Shouldn't the `getError()` method always return a reason?

markstory commented on src/Controller/Component/FormProtectionComponent.php:
```diff
@@ -0,0 +1,162 @@
+     * @param array $params Params
+     * @return \Cake\Http\Response|null
+     */
+    protected function executeCallback(callable $callback, array $params): ?Response
```
Should we require a closure? Users will have access to `Closure::fromCallable()`

markstory commented on src/Controller/Component/SecurityComponent.php:
```diff
@@ -37,6 +37,8 @@
  *
  * @link https://book.cakephp.org/3.0/en/controllers/components/security.html
+ * @deprecated 4.0.0 Use FormProtectionComponent instead, for form tampering protection
+ *   or HttpsEnforceMiddleware to enforce use of HTTPS (SSL) for requests.
```
```suggestion
 *   or HttpsEnforcerMiddleware to enforce use of HTTPS (SSL) for requests.
```

markstory commented on src/Form/FormProtector.php:
```diff
@@ -0,0 +1,605 @@
+ * - Existing fields have not been removed from the form.
+ * - Values of hidden inputs have not been changed.
+ */
+class FormProtector
```
Should `SecurityComponent` also use this class? That would let us ensure compatibility between `FormProtectionComponent` and `SecurityComponent` and maintain only one copy of complex/sensitive code.

markstory commented on src/Form/FormProtector.php:
```diff
@@ -0,0 +1,605 @@
+
+    /**
+     * Parses the field name to create a dot separated name value for use in
+     * field hash. If filename is of form Model[field] or Model.fiedl an array of
```
```suggestion
     * field hash. If filename is of form Model[field] or Model.field an array of
```

markstory commented on src/Form/FormProtector.php:
```diff
@@ -0,0 +1,605 @@
+            return null;
+        }
+
+        $message = '\'%s\' was not found in request data.';
```
```suggestion
        $message = '`%s`' was not found in request data.';
```
Will get formatted better by the development error pages.

markstory commented on src/Form/FormProtector.php:
```diff
@@ -0,0 +1,605 @@
+            return null;
+        }
+        if (!Configure::read('debug') && isset($formData['_Token']['debug'])) {
+            $this->debugMessage = 'Unexpected \'_Token.debug\' found in request data';
```
```suggestion
            $this->debugMessage = 'Unexpected `_Token.debug` found in request data';
```

markstory commented on src/Form/FormProtector.php:
```diff
@@ -0,0 +1,605 @@
+        $expectedUrl = Hash::get($expectedParts, 0);
+        $url = Hash::get($hashParts, 'url');
+        if ($expectedUrl !== $url) {
+            $messages[] = sprintf('URL mismatch in POST data (expected \'%s\' but found \'%s\')', $expectedUrl, $url);
```
```suggestion
            $messages[] = sprintf('URL mismatch in POST data (expected `%s` but found `%s`)', $expectedUrl, $url);
```

If you are on-board with using backticks, I can update all the other error messages and fix tests.

markstory commented on src/View/Helper/SecureFieldTokenTrait.php:
```diff
@@ -1,71 +0,0 @@
- * Provides methods for building token data that is
- * compatible with SecurityComponent.
- */
-trait SecureFieldTokenTrait
```
We should add this removal to the migration guide.

ADmad committed:
Update as per feedback

ADmad committed:
Fix typo

ADmad committed:
Use backticks instead of single quotes in error messages.

ADmad commented:
@markstory Updates done as per feedback. It would be a big chore to update the tests of `SecurityComponent` if it was updated to use the `FormProtector` class :slightly_frowning_face:. 

`SecurityComponent` is deprecated now and should only be used by those upgrading from 3.x, for a short term until they can update their code to use the new `FormProtectionComponent`.

ADmad commented:
"Static Analysis" job failure are unrelated and fixed on `4.x`.

ADmad committed:
Pass exception instance to validation failure callback.

ADmad committed:
Mark FormProtector as internal.
